validator/client/validator.go

```diff
@@ -135,53 +142,74 @@ func (v *validator) SlotDeadline(slot uint64) time.Time {
 // beginning of a new epoch.
 func (v *validator) UpdateAssignments(ctx context.Context, slot uint64) error {
 	// Testing run time for fetching every slot. This is not meant for production!
-	if slot%params.BeaconConfig().SlotsPerEpoch != 0 && v.assignment != nil && false {
+	if slot%params.BeaconConfig().SlotsPerEpoch != 0 && v.assignments != nil && false {
 		// Do nothing if not epoch start AND assignments already exist.
 		return nil
 	}
 
 	ctx, span := trace.StartSpan(ctx, "validator.UpdateAssignments")
 	defer span.End()
-
+	pks := [][]byte{}
+	for i := range v.keys {
+		pks = append(pks, v.keys[i].PublicKey.Marshal())
+	}
 	req := &pb.CommitteeAssignmentsRequest{
 		EpochStart: slot,
-		PublicKey:  [][]byte{v.key.PublicKey.Marshal()},
+		PublicKey:  pks,
 	}
 
 	resp, err := v.validatorClient.CommitteeAssignment(ctx, req)
 	if err != nil {
-		v.assignment = nil // Clear assignments so we know to retry the request.
+		v.assignments = nil // Clear assignments so we know to retry the request.
 		return err
 	}
 
-	v.assignment = resp
+	v.assignments = resp
+	for _, assignment := range v.assignments.Assignment {
+		var proposerSlot uint64
+		var attesterSlot uint64
+		if assignment.IsProposer {
+			proposerSlot = assignment.Slot
+			attesterSlot = assignment.Slot
+		} else {
+			attesterSlot = assignment.Slot
+		}
 
-	lFields := logrus.Fields{
-		"attesterSlot": resp.Assignment[0].Slot - params.BeaconConfig().GenesisSlot,
-		"proposerSlot": "Not proposing",
-		"shard":        resp.Assignment[0].Shard,
-	}
-	if v.assignment.Assignment[0].IsProposer {
-		lFields["proposerSlot"] = resp.Assignment[0].Slot - params.BeaconConfig().GenesisSlot
-	}
+		lFields := logrus.Fields{
+			"attesterSlot": attesterSlot - params.BeaconConfig().GenesisSlot,
+			"proposerSlot": "Not proposing",
+			"shard":        assignment.Shard,
+		}
+		if assignment.IsProposer {
+			lFields["proposerSlot"] = proposerSlot - params.BeaconConfig().GenesisSlot
+		}
+
+		log.WithFields(lFields).Info("Updated validator assignments")
```

Comment:
Is there any way this can be improved? It seems this will log for every assignment when it could log once in some sort of map. 

Imagine we are streaming logs to a database, we don't really want 8 entries for a single update of assignments when running 8 validators.

Reply:
fixed